### DISCUSSION TITLE
Is there already a plugin for highlighting matching paranthesis/brackets?

### DISCUSSION THREAD

**Discussion by bassamsdata** [LazyVim/LazyVim#96] [Q&A] [answered]
I'm looking for an option to **highlight** the matching parenthesis/brackets?

I noticed that a lot of plugins already installed, so I'm wondering if one of those plugins can do that (I already did my search but found none) before installing new plugin?

I was using this plugin `https://github.com/p00f/nvim-ts-rainbow` in my old config but it looks like no longer maintained. 

Any suggestion is appreciated. Thanks

**Comment by al-ce** [comment, marked as answer]
Check out the [new ts-rainbow fork]( https://github.com/mrjones2014/nvim-ts-rainbow) by @mrjones2014. [Reddit thread with the fork announcement](https://www.reddit.com/r/neovim/comments/103x4ux/maintained_fork_of_nvimtsrainbow/)

**Reply by bassamsdata** [reply]
thank you for the answer

**Reply by bassamsdata** [reply]
I'm trying to Highlight match parentheses under cursor? but I'm not been able to.

I think should be enabled by default, right? I'm not sure if the starter LazyVim is disabling that.

Do you have any idea?

**Reply by folke** [reply]
You can either enable the builtin plugin in your `lua/config/lazy.lua` file, or add vim-matchup:

```lua
  {
    "andymass/vim-matchup",
    event = "BufReadPost",
    config = function()
      vim.g.matchup_matchparen_offscreen = { method = "status_manual" }
    end,
  },
```

**Reply by bassamsdata** [reply]
Thank you so much, I'v been reading docs for an hour to activate that hahah :) (I tried all `vim.g.loaded_matchparen = true` and `vim.opt.matchparen` option) but didn't work even though I commented out the `matchparen`. Probably because I was adding those options so there was a confilicts.

Not it worked with just comment out `matchpatern` in `lazy.lua`

Thanks again

**Reply by folke** [reply]
Yeah, that was wrong. Commented those lines in the starter now as well.

**Comment by cgjosephlee** [comment]
https://github.com/mrjones2014/nvim-ts-rainbow is now deprecated, and the author suggested to use another repo https://github.com/HiPhish/nvim-ts-rainbow2/.

I am trying to migrate to the new one, but I have trouble on configuration. Below is my `plugins/custom.lua`:
```lua
return {
  {
    "HiPhish/nvim-ts-rainbow2",
    lazy = true,
  },
  {
    "nvim-treesitter/nvim-treesitter",
    dependencies = { "HiPhish/nvim-ts-rainbow2" },
    opts = function()
    return {
      rainbow = {
        enable = true,
        query = "rainbow-parens",
        strategy = require("ts-rainbow").strategy.global,
      },
    } end,
  },
}
```
and I got errors like `loop or previous error loading module 'ts-rainbow.lib'`, which is similar to https://github.com/HiPhish/nvim-ts-rainbow2/issues/23.

The potential solutions in the thread seems to break the default settings of treesitter in LazyVim. Can someone please provide a correct way to configure this plugin?

**Reply by folke** [reply]
Can you see what happens if you remove this line?

`dependencies = { "HiPhish/nvim-ts-rainbow2" },`

**Reply by cgjosephlee** [reply]
Thanks for your quick reply! Sadly, this does not make any change.

**Reply by dpetka2001** [reply]
I just tried it, but instead of putting it in `opts` and return the options, I put it in the `config` function like mentioned in the post you linked and it works fine. I don't know if there's a more succinct distinction on the difference between `opts` and `config`, because I thought they worked the same in `lazy.nvim`.

**Reply by dpetka2001** [reply]
I tried it putting it in the `opts` like you configured it and get the same error you mentioned. But `config` seems to work just fine.

**Reply by cgjosephlee** [reply]
`config` keyword is already defined in https://github.com/LazyVim/LazyVim/blob/47111bd2db140aac32422752e07db7a1c5b74659/lua/lazyvim/plugins/treesitter.lua#L69

I am afraid that it could break the defaults.

**Reply by dpetka2001** [reply]
Yes it might, I'm not entirely sure about that, so we should wait for a more correct response from @folke to clarify how it should be done. I just mentioned it, because it works if we put the configuration in there.

**Reply by dpetka2001** [reply]
Yes, it does overwrite the default `opts`. Just checked from `TSModuleInfo`. Not sure how to go about this in the `opts` though.

**Reply by jyuan0** [reply]
> Not sure how to go about this in the opts though.

I think it'll work if `opts` is a table, or a function that modifies a table:
```
opts = {
  rainbow = {
    ...
  },
},
```
or
```
opts = function(_, opts)
  opts.rainbow = {
    ...
  }
end,
```

If you define a `config` function, you'll override the default function that LazyVim uses and lose that logic/behavior. IIRC, LazyVim will raise a warning about overridden `config` functions in `checkhealth`.

**Reply by cgjosephlee** [reply]
But the `require` part in `opts` seems to be erroneous.

**Comment by folke** [comment]
This was caused by a custom thing I did for better performance for `treesitter-textobjects`.

I've just changed that implementation.

To use rainbow, do:

```lua
  {
    "nvim-treesitter/nvim-treesitter",
    dependencies = { "HiPhish/nvim-ts-rainbow2" },
    opts = function(_, opts)
      opts.rainbow = {
        enable = true,
        query = "rainbow-parens",
        strategy = require("ts-rainbow").strategy.global,
      }
    end,
  },

**Reply by cgjosephlee** [reply]
Thanks for all your efforts!
